code/datums/components/trader/trader.dm

```diff
@@ -0,0 +1,413 @@
+		return COMPONENT_INCOMPATIBLE
+	var/mob/living/trader = parent
+	if (!trader.ai_controller)
+		return COMPONENT_INCOMPATIBLE
+	if (!trader_data_path)
+		CRASH("Initialised trader component with no trader data.")
+
+	trader_data = new trader_data_path()
+
+	trader.ai_controller.set_blackboard_key(BB_SHOP_SPOT_TYPE, trader_data.shop_spot_type)
+	trader.ai_controller.set_blackboard_key(BB_SHOP_SIGN_TYPE, trader_data.sign_type)
+	trader.ai_controller.set_blackboard_key(BB_SHOP_SOUND, trader_data.sell_sound)
+
+	radial_icons_cache = list(
+		TRADER_RADIAL_BUY = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_buy"),
+		TRADER_RADIAL_SELL = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_sell"),
+		TRADER_RADIAL_TALK = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_talk"),
+		TRADER_RADIAL_LORE = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_lore"),
+		TRADER_RADIAL_DISCUSS_BUY = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_buying"),
+		TRADER_RADIAL_DISCUSS_SELL = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_selling"),
+		TRADER_RADIAL_YES = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_yes"),
+		TRADER_RADIAL_NO = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_no"),
+		TRADER_RADIAL_OUT_OF_STOCK = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_center"),
+	)
+
+	restock_products()
+	renew_item_demands()
+
+/datum/component/trader/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_ATOM_ATTACK_HAND, PROC_REF(on_attack_hand))
+
+/datum/component/trader/UnregisterFromParent()
+	UnregisterSignal(parent, COMSIG_ATOM_ATTACK_HAND)
+
+///If our trader is alive, and the customer left clicks them with an empty hand without combat mode
+/datum/component/trader/proc/on_attack_hand(atom/source, mob/living/carbon/customer)
+	SIGNAL_HANDLER
+	var/mob/living/living_trader = parent
+	if(!can_trade(customer) || customer.combat_mode)
+		return
+	var/list/npc_options = list()
+	if(products.len)
+		npc_options["Buy"] = radial_icons_cache[TRADER_RADIAL_BUY]
+	if(wanted_items.len)
+		npc_options["Sell"] = radial_icons_cache[TRADER_RADIAL_SELL]
+	if(length(trader_data.say_phrases))
+		npc_options["Talk"] = radial_icons_cache[TRADER_RADIAL_TALK]
+	if(!npc_options.len)
+		return
+
+	living_trader.face_atom(customer)
+
+	INVOKE_ASYNC(src, PROC_REF(open_npc_options), customer, npc_options)
+
+	return COMPONENT_CANCEL_ATTACK_CHAIN
+
+/**
+ * Generates a radial of the initial radials of the NPC
+ * Called via asynch, due to the sleep caused by show_radial_menu
+ * Arguments:
+ * * customer - (Mob REF) The mob trying to buy something
+ */
+/datum/component/trader/proc/open_npc_options(mob/living/carbon/customer, list/npc_options)
+	if(!can_trade(customer))
+		return
+	var/npc_result = show_radial_menu(customer, parent, npc_options, custom_check = CALLBACK(src, PROC_REF(check_menu), customer), require_near = TRUE, tooltips = TRUE)
+	switch(npc_result)
+		if("Buy")
```

Comment:
We're already flush with defines but when someone's reusing a string as a key in several places it always strikes me as a good idea to make it a define too, just to prevent any spelling mishaps for future developers.